File: virtual-world-simulator/Fox.cpp
#include "Fox.h"

Fox::Fox(World* world, Point position) :
    Animal{ world, position, 'F', 3, 7 }
{
}

void Fox::update()
{
    int currentX = this->position.x;
    int currentY = this->position.y;
    int newX = currentX;
    int newY = currentY;

    std::vector<Point> possibleMoves;
    Entity*** map = world->getMap();

    for (int i = currentY - 1; i < currentY + 1; i++)
    {
        for (int j = currentX - 1; j < currentX + 1; j++)
        {
			if(i < 0 || i > world->getHeight() - 1 || j < 0 || j > world->getWidth() - 1)
				continue;

            if (i == currentY && j == currentX)
                continue;

            if (map[i][j] == nullptr || map[i][j]->getStrength() <= this->strength)
                possibleMoves.push_back(Point(i, j));
		}
    }
    if(possibleMoves.size() > 0)
    {
        Point newPosition = possibleMoves[Utility::random(0, possibleMoves.size() - 1)];
        this->move(newPosition);
    }
    Entity::update();
}

bool Fox::collision(Entity& entity)
{
    Fox* fox = dynamic_cast<Fox*>(&entity);
    if (fox != nullptr)
    {
        this->breed<Fox>(*fox);
        return false;
    }

    return Animal::collision(entity);
}

File: Java/virtual-world-simulator/src/Entities/Animals/Fox.java
package Entities.Animals;

import App.World;
import Common.Point;
import Entities.Entity;
import GUI.MapElement;

import javax.swing.*;
import java.util.ArrayList;
import java.util.Random;

public class Fox extends Animal
{
    public Fox(World world, Point position)
    {
        super(new ImageIcon("assets/fox.png"), world, position, 'F', 3, 7);
    }

    public boolean Collision(Entity entity)
    {
        if(entity instanceof Fox)
        {
            Fox fox = new Fox(this.world, new Point(-1, -1));
            AddChild(fox);
            return false;
        }

        return super.Collision(entity);
    }

    public void Update()
    {
        int currentX = position.x;
        int currentY = position.y;
        ArrayList<MapElement> possibleMoves = new ArrayList<MapElement>();

        for(int i = currentX - 1; i <= currentX + 1; i++)
        {
            if(i < 0 || i > world.worldHeight - 1)
                continue;

            for(int j = currentY - 1; j <= currentY + 1; j++)
            {
                if(j < 0 || j > world.worldWidth - 1 || (i == currentX && j == currentY))
                    continue;

                if(world.map[i][j].getEntity() == null)
                {
                    possibleMoves.add(world.map[i][j]);
                }
            }
        }

        if(possibleMoves.size() > 0)
        {
            Random rnd = new Random();
            int randomIndex = rnd.nextInt(possibleMoves.size());
            Move(possibleMoves.get(randomIndex).getPosition());
        }
        super.Update();
    }
}
